tachyon/zk/lookup/logup/prover_impl.h

```diff
@@ -0,0 +1,390 @@
+                        chunk[i] = beta + evals.evaluations()[start + i];
+                      }
+                      CHECK(F::BatchInverseInPlaceSerial(chunk));
+                    });
+}
+
+// static
+template <typename Poly, typename Evals>
+template <typename PCS>
+BlindedPolynomial<Poly, Evals> Prover<Poly, Evals>::CreateGrandSumPoly(
+    ProverBase<PCS>* prover, const Evals& m_values,
+    const std::vector<Evals>& compressed_inputs, const Evals& compressed_table,
+    const F& beta) {
+  size_t n = prover->pcs().N();
+
+  // Σ 1/(φᵢ(X))
+  std::vector<F> inputs_log_derivatives(n, F::Zero());
+  std::vector<F> input_log_derivatives(
+      compressed_inputs[0].evaluations().size());
+  for (const Evals& compressed_input : compressed_inputs) {
+    ComputeLogDerivatives(compressed_input, beta, input_log_derivatives);
+
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < n; ++i) {
+      inputs_log_derivatives[i] += input_log_derivatives[i];
+    }
+  }
+
+  // 1 / τ(X)
+  std::vector<F> table_log_derivatives(compressed_table.evaluations().size());
+  ComputeLogDerivatives(compressed_table, beta, table_log_derivatives);
+
+  std::vector<F> grand_sum(n);
+  grand_sum[0] = F::Zero();
+
+  // (Σ 1/φᵢ(X)) - m(X) / τ(X)
+  RowIndex usable_rows = prover->GetUsableRows();
+  std::vector<F> log_derivatives_diff(usable_rows);
+  OPENMP_PARALLEL_FOR(size_t i = 0; i < usable_rows; ++i) {
+    log_derivatives_diff[i] =
+        inputs_log_derivatives[i] - m_values[i] * table_log_derivatives[i];
+    if (i != usable_rows - 1) {
+      grand_sum[i + 1] = log_derivatives_diff[i];
+    }
+  }
+
+// let L(X) = (Σ 1/φᵢ(X)) - m(X) / τ(X)
+// ϕ(ω⁰) = 0
+// ϕ(ω¹) = L(ω⁰)
+// ...
+// ϕ(ω^last) = L(ω⁰) + L(ω¹) + ... + L(ω^{usable_rows - 1})
+#if defined(TACHYON_HAS_OPENMP)
+  size_t active_size = prover->GetUsableRows();
+  size_t thread_nums = static_cast<size_t>(omp_get_max_threads());
+  size_t chunk_size = active_size / thread_nums;
+  if (chunk_size < thread_nums) {
+    chunk_size = 1;
+  }
+  size_t num_chunks = (active_size + chunk_size - 1) / chunk_size;
+
+  std::vector<F> segment_sum(num_chunks, F::Zero());
+
+  OPENMP_PARALLEL_FOR(size_t chunk_idx = 0; chunk_idx < num_chunks;
+                      ++chunk_idx) {
+    size_t start = chunk_idx * chunk_size;
+    size_t end = std::min(start + chunk_size, active_size);
+    for (size_t i = start + 1; i < end; ++i) {
+      grand_sum[i] = grand_sum[i - 1] + log_derivatives_diff[i - 1];
+    }
+    segment_sum[chunk_idx] = grand_sum[end - 1];
+  }
+
+  for (size_t i = 1; i < segment_sum.size(); ++i) {
+    segment_sum[i] += segment_sum[i - 1];
+  }
+
+  OPENMP_PARALLEL_FOR(size_t chunk_idx = 1; chunk_idx < num_chunks;
+                      ++chunk_idx) {
+    size_t start = chunk_idx * chunk_size;
+    size_t end = std::min(start + chunk_size, active_size);
+    F prefix_sum = segment_sum[chunk_idx - 1];
+    for (size_t i = start; i < end; ++i) {
+      grand_sum[i] += prefix_sum;
+    }
+  }
```

Comment:
What about using `base::Parallelize()`?

Reply:
The proof is wrong when I use the base::Parallelize(). I guess it is because the Parallelize gets `num_chunks` with the container's size but in this case we use a `usable_rows`.